frontend_server/lib/server.dart

```diff
@@ -0,0 +1,214 @@
+
+String _usage = '''
+Usage: server [options] [input.dart]
+
+If input dart source code is provided on the command line, then the server 
+compiles it, generates dill file and exits.
+If no input dart source is provided on the command line, server waits for 
+instructions from stdin.
+
+Instructions:
+- compile <input.dart>
+- recompile <boundary-key>
+<path/to/updated/file1.dart>
+<path/to/updated/file2.dart>
+...
+<boundary-key>
+- accept
+- reject
+- quit
+
+Output:
+- result <boundary-key>
+<compiler output>
+<boundary-key> [<output.dill>]
+
+Options:
+${_argParser.usage}
+''';
+
+enum _State { READY_FOR_INSTRUCTION, RECOMPILE_LIST }
+
+/// Actions that every compiler should implement.
+abstract class CompilerInterface {
+  /// Compile given Dart program identified by `filename` with given list of
+  /// `options`.
+  Future<Null> compile(String filename, ArgResults options);
+
+  /// Assuming some Dart program was previously compiled, recompile it again
+  /// taking into account some changed(invalidated) sources.
+  Future<Null> recompileDelta();
+
+  /// Accept results of previous compilation so that next recompilation cycle
+  /// won't recompile sources that were previously reported as changed.
+  void acceptLastDelta();
+
+  /// Reject results of previous compilation. Next recompilation cycle will
+  /// recompile sources indicated as changed.
+  void rejectLastDelta();
+
+  /// This let's compiler know that source file identifed by `uri` was changed.
+  void invalidate(Uri uri);
+}
+
+class _FrontendCompiler implements CompilerInterface {
+  IncrementalKernelGenerator _generator;
+  String _filename;
+  StringSink _outputStream;
+
+  _FrontendCompiler(this._outputStream) {
+    _outputStream ??= stdout;
+  }
+
+  @override
+  Future<Null> compile(String filename, ArgResults options) async {
+    final String boundaryKey = new Uuid().generateV4();
+    _outputStream.writeln("result $boundaryKey");
+    final Uri sdkRoot = _ensureFolderPath(options['sdk-root']);
+    final CompilerOptions compilerOptions = new CompilerOptions()
+      ..sdkRoot = sdkRoot
+      ..strongMode = false
+      ..target = new FlutterTarget(new TargetFlags())
+      ..onError = (CompilationMessage message) {
+        _outputStream.writeln("$message");
+      };
+    Program program;
+    if (options['incremental']) {
+      _generator = await IncrementalKernelGenerator.newInstance(
+          compilerOptions, Uri.base.resolve(filename)
+      );
+      final DeltaProgram deltaProgram = await _generator.computeDelta();
+      program = deltaProgram.newProgram;
+    } else {
+      // TODO(aam): Remove linkedDependencies once platform is directly embedded
+      // into VM snapshot and http://dartbug.com/30111 is fixed.
+      compilerOptions.linkedDependencies = <Uri>[
+        sdkRoot.resolve('platform.dill')
+      ];
+      program =
+          await kernelForProgram(Uri.base.resolve(filename), compilerOptions);
```

Comment:
no need for the newline here. just put it all on one line. The 4 characters saved aren't worth the readability hit.